File: mod/io/echo.ts
//:echo
const echo = ((): returnAction => {
    if (DEBUG) {
        let varname = args.split(' ')[0]
        let space = longest - varname.length
        space = space < 0 ? 0 : space
        console.log(`[${varname}]${new Array(space).fill(' ').join('')}| ` + (str_memory[varname] ?? num_memory[varname] ?? '(null)'))
    } else {
        console.log(str_memory[args.split(' ')[0]] ?? '(null)')
    }

    return returnAction.PEACEFUL
})

File: nim/mod/io/echo.nim
template mod_io_echo() =
    proc d_echo(): returnAction =
        var str = args
        for mt in str.findAll(re"%{[^} ]+}"):
            var name = mt.substr(2, mt.len - 2)
            var res = (if str_memory.hasKey(name): str_memory[name] 
                elif num_memory.hasKey(name): $(num_memory[name])
                else: "(null)")
            str = str.replace(mt, res)
        
        echo str
        return returnAction.PEACEFUL

    proc printf(): returnAction =
        var str = args
        for mt in str.findAll(re"%{[^} ]+}"):
            var name = mt.substr(2, mt.len - 2)
            var res = (if str_memory.hasKey(name): str_memory[name] 
                elif num_memory.hasKey(name): $(num_memory[name])
                else: "(null)")
            str = str.replace(mt, res)
        
        stdout.write str
        return returnAction.PEACEFUL
    
    internals["echo"] = d_echo
    internals["printf"] = printf

export mod_io_echo